internal/configs/import.go

```diff
@@ -76,3 +90,76 @@ var importBlockSchema = &hcl.BodySchema{
 		},
 	},
 }
+
+// parseResourceInstanceFromExpression takes an arbitrary expression
+// representing a resource instance, and parses out the static ConfigResource
+// skipping an variable index expressions. This is used to connect an import
+// block's "to" to the configuration address before the full instance
+// expressions are evaluated.
+func parseConfigResourceFromExpression(expr hcl.Expression) (addrs.ConfigResource, tfdiags.Diagnostics) {
+	traversal, hcdiags := exprToResourceTraversal(expr)
+	if hcdiags.HasErrors() {
+		return addrs.ConfigResource{}, tfdiags.Diagnostics(nil).Append(hcdiags)
+	}
+
+	addr, diags := addrs.ParseAbsResourceInstance(traversal)
+	if diags.HasErrors() {
+		return addrs.ConfigResource{}, diags
+	}
+
+	return addr.ConfigResource(), diags
+}
+
+// exprToResourceTraversal is used to parse the import block's to expression,
+// which must be a resource instance, but may contain limited variables with
+// index expressions. Since we only need the ConfigResource to connect the
+// import to the configuration, we skip any index expressions.
+func exprToResourceTraversal(expr hcl.Expression) (hcl.Traversal, hcl.Diagnostics) {
+	var trav hcl.Traversal
+	var diags hcl.Diagnostics
+
+	switch e := expr.(type) {
+	case *hclsyntax.RelativeTraversalExpr:
+		t, d := exprToResourceTraversal(e.Source)
+		diags = diags.Extend(d)
+		trav = append(trav, t...)
+		trav = append(trav, e.Traversal...)
+
+	case *hclsyntax.ScopeTraversalExpr:
+		// a static reference, we can just append the traversal
+		trav = append(trav, e.Traversal...)
+
+	case *hclsyntax.IndexExpr:
+		// Get the collection from the index expression, we don't need the
+		// index for a ConfigResource
+		t, d := exprToResourceTraversal(e.Collection)
+		diags = diags.Extend(d)
+		if diags.HasErrors() {
+			return nil, diags
+		}
+		trav = append(trav, t...)
+
+	default:
+		// if we don't recognise the expression type (which means we are likely
+		// dealing with a test mock), try and interpret this as an absolute
+		// traversal
+		t, d := hcl.AbsTraversalForExpr(e)
```

Comment:
I think an `import` block declared in a `.tf.json` file will end up in this case too, and `hcl.AbsTraversalForExpr` in that context has some interesting consequences.

In particular, `hcl.AbsTraversalForExpr` on a JSON `hcl.Expression` will expect a JSON string like `"aws_instance.foo"`, whereas normal expression evaluation would expect `"${aws_instance.foo}"`.

We can recognize whether a `hcl.Expression` originated in a JSON document by calling [`json.IsJSONExpression`](https://pkg.go.dev/github.com/hashicorp/hcl/v2/json#IsJSONExpression), which is really just a thin wrapper around a type assertion of the `json` package's unexported concrete expression type.

I think actually _handling_ a JSON expression once detected will probably require adding some new features to HCL's JSON package which we've considered a few times now but have so far always ended up finding an alternative:

The high-level idea is to expose a set of three methods in the package which take an expression for which `IsJSONExpression` would return `true` (panicking for any other expression, since that'd be a caller bug), and interprets the JSON expression in one of the following three ways:

1. As an evaluable expression, returning a `hclsyntax.Expression`:
    - a JSON object becomes a `hclsyntax.ObjectConsExpr` with the same function recursively applied to all of the nested values.
    - a JSON array becomes a `hclsyntax.TupleConsExpr` with the same function recursively applied to all of the nested values.
    - a JSON string gets parsed with `hclsyntax.ParseTemplate` and the result is returned. This should use equivalent logic to what the current `expr.Value` implementation does when given a non-nil EvalContext and evaluating a JSON string.
    - everything else (I think?) just gets trivially evaluated to a `cty.Value` and wrapped in a `hclsyntax.LiteralValueExpr`
2. As a static-analysis-friendly expression, also returning an `hcl.Expression` but with one detail different than the previous item:
    - a JSON string gets parsed with `hclsyntax.ParseExpression` instead, and the result is returned.
3. As a literal, returning a `cty.Value` directly: we can technically already do this with` expr.Value(nil)`, but I'd propose adding a function specifically for it to complete this set and also make that possibility a little more evident in the package API. (Terraform uses this to ensure that `.tfvars.json` files are not subject to HCL template parsing/evaluation, for example.)

The three separate functions here reflect the three different ways a JSON value can be interpreted under the rules of the JSON variant of HCL. I think that would then allow some existing `json` package logic to be reimplemented in terms of the above three functions:
- `hcl.AbsTraversalForExpr`, `hcl.RelTraversalForExpr`, `hcl.ExprList`, `hcl.ExprMap`, and `hcl.ExprCall` with a JSON expression would interpret the given expression as case 2 above, and then call the same outer function again on the resulting `hclsyntax.Expression`.
- `expr.Value` for a JSON expression would interpret the given expression either as case 1 or case 3 depending on whether the given `EvalContext` is `nil`, and then call `Value` on the resulting `hclsyntax.Expression`.

But most importantly for what you were trying to do here, you'd also be able to directly call function 2 in this `exprToResourceTraversal` function if `json.IsJSONExpression` returns true, and then analyze the returned `hclsyntax.Expression` instead. This would, I think, effectively bridge the gap we've encountered many times now where the JSON syntax _is_ implemented in terms of `hclsyntax` but the translations it does are hidden in the unexported implementation. If we later also add functionality to `hclwrite` to take an `hclsyntax.Expression` and return an equivalent [`hclwrite.Expression`](https://pkg.go.dev/github.com/hashicorp/hcl/v2/hclwrite#Expression) then we'd have the building blocks required for a JSON-syntax-to-native-syntax translation tool, too.

If you come up with a more straightforward way to get there then of course it would be very reasonable not to take on a big reorganization of HCL's JSON syntax implementation as part of this project, but I offer the above in case it's useful since I expect the effort of implementing it will pay off in various other contexts too.

Reply:
Thanks for jumping into this early @apparentlymart! I was about to start validating json input (which I unfortunately left til the end of course 🙄), and this gives me a good launching point when I find it doesn't work 😉

Reply:
On further reflection I think I'd actually choose a different variation on function 3, for consistency with the others:

3. The same as cases 1 and 2 except that JSON strings get evaluated to their equivalent `cty.StringVal` and returned as `hclsyntax.LiteralValueExpr`.

This means that all three functions take a JSON expression and return a `hclsyntax.Expression`, and the only difference between them is what they do when encountering JSON strings. That string treatment is the essence of what makes string processing ambiguous in the JSON syntax and so I think it makes sense that it would be the only difference between these cases.

We could potentially still offer the variation that returns a `cty.Value` directly, since in that case it's guaranteed that the result will only include `cty.ObjectConsExpr`, `cty.TupleConsExpr`, or `cty.LiteralValueExpr` nodes in its entire subtree and so it's trivial to just call `Value(nil)` on the result to get that effect, but having three functions that all share an identical signature `func (hcl.Expression) hclsyntax.Expression` would make this more composable.